File: results/0.25/DNS/dns-gpt-4-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

// DNS Header structure
typedef struct {
    uint16_t id;
    uint8_t qr;
    uint8_t opcode;
    uint8_t aa;
    uint8_t tc;
    uint8_t rd;
    uint8_t ra;
    uint8_t z;
    uint8_t rcode;
    uint16_t qdcount;
    uint16_t ancount;
    uint16_t nscount;
    uint16_t arcount;
} DNSHeader;

// DNS Question structure
typedef struct {
    HBytes *qname;
    uint16_t qtype;
    uint16_t qclass;
} DNSQuestion;

// DNS Resource Record structure
typedef struct {
    HBytes *name;
    uint16_t type;
    uint16_t class;
    uint32_t ttl;
    uint16_t rdlength;
    HBytes *rdata;
} DNSResourceRecord;

// Function to parse DNS Header
HParser *parse_dns_header() {
    return h_sequence(
        h_uint16(), // ID
        h_bits(1, false), // QR
        h_bits(4, false), // Opcode
        h_bits(1, false), // AA
        h_bits(1, false), // TC
        h_bits(1, false), // RD
        h_bits(1, false), // RA
        h_bits(3, false), // Z
        h_bits(4, false), // RCODE
        h_uint16(), // QDCOUNT
        h_uint16(), // ANCOUNT
        h_uint16(), // NSCOUNT
        h_uint16(), // ARCOUNT
        NULL
    );
}

// Function to parse DNS Question
HParser *parse_dns_question() {
    return h_sequence(
        h_length_value(h_uint8(), h_uint8()), // QNAME
        h_uint16(), // QTYPE
        h_uint16(), // QCLASS
        NULL
    );
}

// Function to parse DNS Resource Record
HParser *parse_dns_rr() {
    return h_sequence(
        h_length_value(h_uint8(), h_uint8()), // NAME
        h_uint16(), // TYPE
        h_uint16(), // CLASS
        h_uint32(), // TTL
        h_length_value(h_uint16(), h_uint8()), // RDLENGTH and RDATA
        NULL
    );
}

// Function to parse entire DNS message
HParser *parse_dns_message() {
    HParser *header = parse_dns_header();
    HParser *question = parse_dns_question();
    HParser *answer = parse_dns_rr();
    HParser *authority = parse_dns_rr();
    HParser *additional = parse_dns_rr();

    return h_sequence(
        header,
        h_many(question),
        h_many(answer),
        h_many(authority),
        h_many(additional),
        NULL
    );
}

int main(int argc, char **argv) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <dns_message_file>\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        perror("Failed to open file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    fseek(file, 0, SEEK_SET);

    uint8_t *data = malloc(length);
    if (!data) {
        perror("Failed to allocate memory");
        fclose(file);
        return 1;
    }

    fread(data, 1, length, file);
    fclose(file);

    HParser *dns_parser = parse_dns_message();
    HParseResult *result = h_parse(dns_parser, data, length);
    if (result) {
        printf("DNS message parsed successfully.\n");
    } else {
        printf("Failed to parse DNS message.\n");
    }

    h_parse_result_free(result);
    free(data);

    return 0;
}

File: results/0.25/DNS/dns-gpt-4-turbo.rs
use nom::{
    bytes::complete::{take, take_while},
    number::complete::{be_u16, be_u32, be_u8},
    sequence::tuple,
    IResult,
};
use std::env;
use std::fs::File;
use std::io::{self, Read};

#[derive(Debug)]
struct DNSHeader {
    id: u16,
    qr: u8,
    opcode: u8,
    aa: u8,
    tc: u8,
    rd: u8,
    ra: u8,
    z: u8,
    ad: u8,
    cd: u8,
    rcode: u8,
    qdcount: u16,
    ancount: u16,
    nscount: u16,
    arcount: u16,
}

#[derive(Debug)]
struct Question {
    qname: Vec<u8>,
    qtype: u16,
    qclass: u16,
}

#[derive(Debug)]
struct ResourceRecord {
    name: Vec<u8>,
    rtype: u16,
    rclass: u16,
    ttl: u32,
    rdlength: u16,
    rdata: Vec<u8>,
}

#[derive(Debug)]
struct DNSPacket {
    header: DNSHeader,
    questions: Vec<Question>,
    answers: Vec<ResourceRecord>,
    authorities: Vec<ResourceRecord>,
    additionals: Vec<ResourceRecord>,
}

fn parse_dns_header(input: &[u8]) -> IResult<&[u8], DNSHeader> {
    let (input, (id, flags, qdcount, ancount, nscount, arcount)) = tuple((be_u16, be_u16, be_u16, be_u16, be_u16, be_u16))(input)?;
    let qr = (flags >> 15) & 0x1;
    let opcode = (flags >> 11) & 0xF;
    let aa = (flags >> 10) & 0x1;
    let tc = (flags >> 9) & 0x1;
    let rd = (flags >> 8) & 0x1;
    let ra = (flags >> 7) & 0x1;
    let z = (flags >> 6) & 0x1;
    let ad = (flags >> 5) & 0x1;
    let cd = (flags >> 4) & 0x1;
    let rcode = flags & 0xF;

    Ok((
        input,
        DNSHeader {
            id,
            qr,
            opcode,
            aa,
            tc,
            rd,
            ra,
            z,
            ad,
            cd,
            rcode,
            qdcount,
            ancount,
            nscount,
            arcount,
        },
    ))
}

fn parse_dns_question(input: &[u8]) -> IResult<&[u8], Question> {
    let (input, qname) = take_while(|b: u8| b != 0)(input)?;
    let (input, _) = take(1usize)(input)?; // null byte of qname
    let (input, (qtype, qclass)) = tuple((be_u16, be_u16))(input)?;

    Ok((
        input,
        Question {
            qname: qname.to_vec(),
            qtype,
            qclass,
        },
    ))
}

fn parse_dns_rr(input: &[u8]) -> IResult<&[u8], ResourceRecord> {
    let (input, name) = take_while(|b: u8| b != 0)(input)?;
    let (input, _) = take(1usize)(input)?; // null byte of name
    let (input, (rtype, rclass, ttl, rdlength)) = tuple((be_u16, be_u16, be_u32, be_u16))(input)?;
    let (input, rdata) = take(rdlength)(input)?;

    Ok((
        input,
        ResourceRecord {
            name: name.to_vec(),
            rtype,
            rclass,
            ttl,
            rdlength,
            rdata: rdata.to_vec(),
        },
    ))
}

fn parse_dns_packet(input: &[u8]) -> IResult<&[u8], DNSPacket> {
    let (input, header) = parse_dns_header(input)?;
    let mut input = input;
    let mut questions = Vec::new();
    for _ in 0..header.qdcount {
        let (new_input, question) = parse_dns_question(input)?;
        questions.push(question);
        input = new_input;
    }

    let mut answers = Vec::new();
    for _ in 0..header.ancount {
        let (new_input, answer) = parse_dns_rr(input)?;
        answers.push(answer);
        input = new_input;
    }

    let mut authorities = Vec::new();
    for _ in 0..header.nscount {
        let (new_input, authority) = parse_dns_rr(input)?;
        authorities.push(authority);
        input = new_input;
    }

    let mut additionals = Vec::new();
    for _ in 0..header.arcount {
        let (new_input, additional) = parse_dns_rr(input)?;
        additionals.push(additional);
        input = new_input;
    }

    Ok((
        input,
        DNSPacket {
            header,
            questions,
            answers,
            authorities,
            additionals,
        },
    ))
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <file>", args[0]);
        return Ok(());
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_dns_packet(&buffer) {
        Ok((_, packet)) => println!("{:#?}", packet),
        Err(e) => eprintln!("Failed to parse DNS packet: {:?}", e),
    }

    Ok(())
}